### ISSUE TITLE
Problem with BDI Step Mode

### ISSUE THREAD

**Issue by osrf-migration** [osrf/drcsim#325]
**Original report ([archived issue](https://osrf-migration.github.io/drcsim-gh-pages/#!/osrf/drcsim/issues/325)) by Richard Stilson (Bitbucket: [GoldFinger90210](https://bitbucket.org/%7Bb0e3fea6-84dd-4d6d-8a30-6e1b7ed3b02b%7D/)).**

----------------------------------------
I have found that the drcsim robot instantly freezes when I enter BDI step control mode under certain circumstances. First, the circumstances in which it works successfully
* If I start with zero heading, do dynamic walking, and either maintain the zero heading, or turn during dynamic walking then turn again back to a zero heading, I can reenter step mode (qual 1 fits this pattern)
* If I use step mode to turn to a different heading, then do dynamic walking along that same heading, then stop, I can reenter step mode (ex. step to .78 heading, begin dynamic walking at .78 heading, then stop and reenter stepping mode still at .78 heading)
* If I use step mode to turn to a different heading, then do dynamic walking, in which I turn, but then turn back to end up at the same heading at which I last finished step mode, I can reenter step mode (ex. step to .78 rad heading, begin dynamic walking, turn to 1.57 heading, then turn back to .78 heading, then stop and reenter stepping mode)

However, if I do the following, the robot instantly freezes and falls over
* Start with stepping and finish on any heading, do dynamic walking and turn to a different heading, then attempt to step
*  Ex.  Start with 0 heading, dynamic walk to .78 heading, stop and attempt stepping with instant fail
*  Ex. Start by stepping to .78 heading, dynamic walk to 0 heading, stop and attempt stepping with instant fail
*  Ex. Start directly into dynamic walking, turn to any non-zero heading, stop and attempt stepping with instant fail

It appears to me that the stepping routine remembers its previous heading, and the robot can only reenter step mode if it is at the same heading as it was when it previously ended step mode. This also seems to apply even when the robot never entered step mode after starting the sim

When this error occurs, there are no warning message from the AtlasSimInterfaceState error flag, or the status flags for behavior feedback, walk feedback, or step feedback. The behavior feedback flag simply goes to 256 instantly. I have examined the step data I command and it is perfectly sensible in each case that results in falling. Note: I am able to reenter step mode at any heading. It just has to be a heading I previously achieved by rotating in step mode. Even if the steps were not well positioned, I would expect the robot to attempt the step and fail, rather than just instantly going stiff and falling.

At this point the only work around I can see is to ensure that if I try to enter step mode, my heading is the same as the last time I was in step mode, or if this is the first time in step mode, my heading must be zero.
Obviously I'd like to be able to enter step mode at an arbitrary heading after dynamic walking

Hopefully this can be fixed easily, since the nature of the error is so specific

I have tested this with the latest version of the drcsim software released today.

**Comment by osrf-migration** [comment]
**Original comment by Richard Stilson (Bitbucket: [GoldFinger90210](https://bitbucket.org/%7Bb0e3fea6-84dd-4d6d-8a30-6e1b7ed3b02b%7D/)).**

----------------------------------------
TL; DR

I think the robot remembers its previous heading from the last time it was in step mode, and can only go back into step mode if it is at the same heading. If the heading changes during dynamic walking, then an attempt to reenter step mode causes the robot to simply freeze and fall over.

**Comment by osrf-migration** [comment]
**Original comment by John Hsu (Bitbucket: [hsu](https://bitbucket.org/%7B0a186eae-abf0-4514-a951-23db5eccc286%7D/), GitHub: [hsu](https://github.com/hsu)).**

----------------------------------------
duplicate of issue #311

**Comment by osrf-migration** [comment]
**Original comment by David Conner (Bitbucket: [dcconner](https://bitbucket.org/%7B4375dc4a-9371-4c51-973e-1f1d045590c5%7D/)).**

----------------------------------------
We don't have time to test the issue 311 pull by building from source, but would be willing if the updates are in drcsim-nightly.

The last I heard, you were punting on the merge request.. We would like to have working step control if possible.

**Comment by osrf-migration** [comment]
**Original comment by John Hsu (Bitbucket: [hsu](https://bitbucket.org/%7B0a186eae-abf0-4514-a951-23db5eccc286%7D/), GitHub: [hsu](https://github.com/hsu)).**

----------------------------------------

* changed state from "new" to "duplicate"

Duplicate of #311.

**Comment by osrf-migration** [comment]
**Original comment by David Conner (Bitbucket: [dcconner](https://bitbucket.org/%7B4375dc4a-9371-4c51-973e-1f1d045590c5%7D/)).**

----------------------------------------
Future comments will be addressed in pull request 351, which addresses this same issue
